### DISCUSSION TITLE
TextArea Cursor/Caret Symbol — Change from Block (`█`) to Beam (`▏`)

### DISCUSSION THREAD

**Discussion by Humboldt-Penguin** [Textualize/textual#4568] [Q&A] [answered]
Hi, hope everyone is doing alright. 

**My Question** — Can you can make the `TextArea` cursor/caret appear as a vertical beam (`▏`) or underline (`_`), as opposed to the default block character (`█`)? 

Hacky solutions are okay (e.g. getting the cursor position and manually rendering the desired caret character). I'd also be willing to try a PR as well if someone can point to some ideas. 

---

**Previous Attempts** — I brute force guessed various properties for `.text-area--cursor`, but none of them were valid. 

[Searching the source code](https://github.com/search?q=repo%3ATextualize%2Ftextual+%E2%96%88&type=code), the block character likely originates from [src/textual/_border.py](https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/_border.py#L86) or [src/textual/widgets/_rule.py](https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/widgets/_rule.py#L51), but I can't find the code that actually renders it to screen. 

---

**Minimal Code** — If you want to tinker:

```python
from textual.app import App, ComposeResult
from textual.widgets import TextArea


class TextAreaExample(App):
    
    # CSS = """\
    # .text-area--cursor {
    #     /* If you want to try CSS */
    # }
    # """


    def compose(self) -> ComposeResult:
        default_text = 'This is a basic text editor.'
        text_area = TextArea(default_text)
        # Note: I don't think the `text_area` object has any attribute to define cursor symbol. 
        yield text_area


if __name__ == "__main__":
    app = TextAreaExample()
    app.run()
```

--- 

**Diagnostics —**

```
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.63.4 |
| Rich    | 13.7.1 |

## Python

| Name           | Value                                             |
|----------------|---------------------------------------------------|
| Version        | 3.12.3                                            |
| Implementation | CPython                                           |
| Compiler       | MSC v.1938 64 bit (AMD64)                         |
| Executable     | C:\Users\Eris\mambaforge\envs\textual1\python.exe |

## Operating System

| Name    | Value      |
|---------|------------|
| System  | Windows    |
| Release | 10         |
| Version | 10.0.19045 |

## Terminal

| Name                 | Value            |
|----------------------|------------------|
| Terminal Application | Windows Terminal |
| TERM                 | *Not set*        |
| COLORTERM            | *Not set*        |
| FORCE_COLOR          | *Not set*        |
| NO_COLOR             | *Not set*        |

## Rich Console options

| Name           | Value               |
|----------------|---------------------|
| size           | width=99, height=53 |
| legacy_windows | False               |
| min_width      | 1                   |
| max_width      | 99                  |
| is_terminal    | True                |
| encoding       | utf-8               |
| max_height     | 53                  |
| justify        | None                |
| overflow       | None                |
| no_wrap        | False               |
| highlight      | None                |
| markup         | None                |
| height         | None                |
```

For context, I'm working on a FOSS/self-hosted alternative to paid/proprietary apps like Workflowy/Dynalist in the "infinite outliner" niche.

Thank you for your time!

**Comment by davep** [comment, marked as answer]
> [Searching the source code](https://github.com/search?q=repo%3ATextualize%2Ftextual+%E2%96%88&type=code), the block character likely originates from [src/textual/_border.py](https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/_border.py#L86) or [src/textual/widgets/_rule.py](https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/widgets/_rule.py#L51), but I can't find the code that actually renders it to screen.

The cursor for `TextArea` is [handled by the source for `TextArea`](https://github.com/Textualize/textual/blob/main/src/textual/widgets/_text_area.py), not sure how you got to those files from there.

The cursor isn't a character, [it's a changing of the background colour of the cell the cursor is on](https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/widgets/_text_area.py#L122-L141). If the cursor *was* a character you wouldn't be able to see the character it was on.

**Reply by allefeld** [reply]
Why not use the regular terminal text cursor, with the advantage that it is styleable, especially in modern terminal emulators? I find a blinking box very ugly.
